server/internal/infrastructure/gcp/cloudtasks.go

```diff
@@ -2,46 +2,90 @@ package gcp
 
 import (
 	"context"
+	"encoding/json"
 
+	cloudtasks "cloud.google.com/go/cloudtasks/apiv2"
 	"github.com/reearth/reearth-cms/server/internal/usecase/gateway"
 	"github.com/reearth/reearth-cms/server/pkg/id"
+	"github.com/reearth/reearth-cms/server/pkg/id/idx"
 	"github.com/reearth/reearth-cms/server/pkg/task"
+	"google.golang.org/api/option"
+	taskspb "google.golang.org/genproto/googleapis/cloud/tasks/v2"
 )
 
-// Queue tasks care of queuing and Tasks will be added to it
-type Queue struct {
+type TaskRunner struct {
+	queuePath     string
+	subscriberURL string
+	credFilePath  string
 }
 
-type RequestType string
+func NewTaskRunner(c *CloudTasksConfig, opts ...TaskRunnerOption) (gateway.TaskRunner, error) {
+	opts2 := defaultTaskRunnerOptions()
+	for _, o := range opts {
+		o.Apply(opts2)
+	}
 
-const (
-	HTTP = RequestType("http")
-)
+	qURL, err := c.buildQueueUrl()
+	if err != nil {
+		return nil, err
+	}
 
-// Task is request for subscriber
-type Task struct {
-	reqType RequestType
-	payload any
+	return &TaskRunner{
+		queuePath:     qURL,
+		subscriberURL: c.SubscriberURL,
+		credFilePath:  opts2.credFilePath,
+	}, nil
 }
 
-type Config struct {
-	GCPProject string
-	GCPRegion  string
-	QueueName  string
-}
+// Run implements gateway.TaskRunner
+func (t *TaskRunner) Run(ctx context.Context, p task.Payload) (id.TaskID, error) {
 
-func NewTaskRunner(c *Config) gateway.TaskRunner {
-	// TODO: convert config to Queue struct
-	return &Queue{}
-}
+	client, closeFn, err := t.client(ctx)
+	if err != nil {
+		return idx.ID[id.Task]{}, err
+	}
+	defer closeFn()
+
+	bPayload, err := json.Marshal(p.DecompressAsset.Payload())
+	if err != nil {
+		return idx.ID[id.Task]{}, err
+	}
+	req := t.buildRequest(t.subscriberURL, bPayload)
+
+	_, err = client.CreateTask(ctx, req)
+	if err != nil {
+		return idx.ID[id.Task]{}, err
+	}
 
-func NewTask() *Task {
 	//TODO: impl here
-	return &Task{}
+	return idx.ID[id.Task]{}, nil
 }
 
-// Run implements gateway.TaskRunner
-func (*Queue) Run(ctx context.Context, p task.Payload) (id.TaskID, error) {
-	//TODO: implement login to add task to the queue with payload
-	panic("unimplemented")
+func (t *TaskRunner) client(ctx context.Context) (*cloudtasks.Client, func(), error) {
+	var c *cloudtasks.Client
+	var err error
+	if t.credFilePath == "" {
+		c, err = cloudtasks.NewClient(ctx)
+	} else {
+		c, err = cloudtasks.NewClient(ctx, option.WithCredentialsFile(t.credFilePath))
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+	return c, func() { c.Close() }, nil
```

Comment:
Though the Cloudtasks client needs to receive `context` which should be limited in request scope when it's initialized, should I still initialize only once at the constructor?
I also thought it shouldn't be initialized every time, but since it receives `context` I chose this way. What do you think? 😄

Reply:
Additionally, though cloudtasks Client has a `Close` function which seems to be called after the request has been sent, do we still avoid to initialize it every time? 🤔

Reply:
@rot1024 Change with [refactor client init](https://github.com/reearth/reearth-cms/pull/127/commits/a1f7b13c0c53ce3c51acc1f9a2848badbfe20768) and [refactor](https://github.com/reearth/reearth-cms/pull/127/commits/6a9e8e8c9528826c966a4661e7adb6b0bf22ac1e). 
The reason why I designed that constructor can receive the `*cloudtasks.Client` is to allow the test package to pass mocked client. 
Since ours doesn't have the package wichi is like `registry`, I assume this is the only way to inject dependencies from outside.